openedx/core/djangoapps/cache_toolbox/middleware.py

```diff
@@ -141,9 +142,21 @@ def _verify_session_auth(self, request):
         auto_auth_enabled = settings.FEATURES.get('AUTOMATIC_AUTH_FOR_TESTING', False)
         if not auto_auth_enabled and hasattr(request.user, 'get_session_auth_hash'):
             session_hash = request.session.get(HASH_SESSION_KEY)
-            if not (session_hash and constant_time_compare(session_hash, request.user.get_session_auth_hash())):
-                # The session hash has changed due to a password
-                # change. Log the user out.
-                request.session.flush()
-                request.user = AnonymousUser()
-                _mark_cookie_for_deletion(request)
+            session_hash_verified = session_hash and constant_time_compare(
+                session_hash, request.user.get_session_auth_hash())
+            if session_hash_verified:
+                set_custom_attribute('CacheBackedAuthenticationMiddleware.verify.alg', settings.DEFAULT_HASHING_ALGORITHM)
+            else:
+                if not (
+                    session_hash and
+                    hasattr(request.user, '_legacy_get_session_auth_hash') and
+                    # pylint: disable: protected-access
+                    constant_time_compare(session_hash, request.user._legacy_get_session_auth_hash())
+                ):
+                    # The session hash has changed due to a password
+                    # change. Log the user out.
+                    request.session.flush()
+                    request.user = AnonymousUser()
+                    _mark_cookie_for_deletion(request)
+                else:
+                    set_custom_attribute('CacheBackedAuthenticationMiddleware.verify.alg', "sha1")
```

Comment:
Nit: 
1. You can decide on this one, but I like keeping my monitoring with the code, rather than creating additional cases. That said, you can do what you wish here.
2. Not sure if it would be useful to have the result of `hasattr(request.user, '_legacy_get_session_auth_hash')` in a custom attribute?
```suggestion
                    # temporary monitoring for failed legacy session check
                    set_custom_attribute('failed_legacy_session_verification', True)
```